### ISSUE TITLE
EmptyLineSeparator does not catch missing line separator before field declaration

### ISSUE THREAD

**Issue by peterdemaeyer** [checkstyle/checkstyle#11286]
The documentation of EmptyLineSeparator says "Checks for empty line separators before package, all import declarations, **fields**, constructors, methods, nested classes, static initializers and instance initializers."
In particular, it doesn't work for fields, see example below.
Since a field is also known as an instance variable, I expect the VARIABLE_DEF to catch it, but it doesn't.

I have read check documentation: https://checkstyle.sourceforge.io/config_whitespace.html#EmptyLineSeparator
https://checkstyle.sourceforge.io/checks/whitespace/emptylineseparator.html#EmptyLineSeparator
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
peter@pangolin:~/Downloads/checkstyle$ javac Test.java 

peter@pangolin:~/Downloads/checkstyle$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
	"-//Puppy Crawl//DTD Check Configuration 1.3//EN"
	"http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
	<module name="TreeWalker">
		<module name="EmptyLineSeparator"/>
	</module>
</module>

peter@pangolin:~/Downloads/checkstyle$ cat Test.java 
class Test {
	int i = 0; // Checkstyle doesn't catch missing line break before this line. Should be caught by VARIABLE_DEF?
	Test() { // Checkstyle DOES catch missing line break before this line. Caught by CTOR_DEF.
	}
}

peter@pangolin:~/Downloads/checkstyle$ RUN_LOCALE="-Duser.language=en -Duser.country=US"

peter@pangolin:~/Downloads/checkstyle$ java $RUN_LOCALE -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/peter/Downloads/checkstyle/Test.java:3:9: 'CTOR_DEF' should be separated from previous line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 1 errors.
```

---
**Expected: Checkstyle reports 2 violations, one for CTOR_DEF and one for VARIABLE_DEF.**
**Actual: Checkstyle reports 1 violation, for CTOR_DEF.**
---

To give an idea of the code style I want to enforce:

```java
class AlwaysNewlineAfterClassDeclaration {

    int field = 0; // Newline before is required.
}

class AlsoInAnymousInnerClasses {

    Supplier<String> supplier = new Supplier<>() {

        public String get() { // Newline before is required.
        }
    }
} // Newline before is prohibited.

class UnlessEmpty {
} // Newline before is prohibited.
```

So I want to be able to configure the following violations:

```java
class AlwaysNewlineAfterClassDeclaration {
    int field = 0; // Violation.
}

class AlsoInAnymousInnerClasses {

    Supplier<String> supplier = new Supplier<>() {
        public String get() { // Violation.
        }
    }

} // Violation, out of scope of this issue.

class UnlessEmpty {

} // Violation, out of scope of this issue.
```

**Comment by peterdemaeyer** [comment]
`CLASS_DEF` and `VARIABLE_DEF` are in the list of tokens, shouldn't the check just work between the two?
The way I see it, `class AlwaysNewlineAfterClassDeclaration {` is the class definition, and `int field = 0` in my aforementioned example is the variable definition, so why doesn't the check require a newline between both?

**Comment by nmancus1** [comment]
It seems like `EmptyLineSeparator` does not check previous line for enclosing type declaration at all:


```
$ javac Test.java

$ cat Test.java
class Test0 {
    final int myInt = 5;
}

class Test1 {
    Test1() {}
}

class Test2 {
    void m() {}
}

class Test3 {
    class nestedClass{}
}

class Test4 {
    static {}
}

class Test5 {
    {}
}
}

$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="EmptyLineSeparator"/>
    </module>
</module>

$ java -jar checkstyle-9.3-all.jar -c config.xml Test.java
Starting audit...
Audit done.


```

I think that issue is a bit more nuanced, though; I believe the intention of this check is to verify empty line separators between "parallel" elements  (i.e. separate class members). Additionally, I have seen that many projects that do use empty lines to separate class members do not use this style for the first class member.

@strkkk what do you think about making this the default behavior for this check?

**Comment by strkkk** [comment]
As for me, this check has plenty of properties, I think new check seems a better way

**Comment by peterdemaeyer** [comment]
I'm not convinced a new check would be better, because it would add complexity whilst addressing largely the same concern as the existing check: to address line separators before/between/after class members. By changing the existing code, I can take the opportunity to make the existing code better.

To explain my rationale, consider the first line of documentation:

> Checks for empty line separators before package, all import declarations, fields, constructors, methods, nested classes, static initializers and instance initializers.

Note that it explicitly says _"before ..."_, which makes perfect sense to me. I imagine that this was the requirement that was written.

However, looking at the implementation, some of the elegance was taken out because the implementation does not check line separators _before_ tokens, but rather _after_ them. I suspect this may have resulted in the first "ATTENTION" in the documentation, which is much harder to a user to understand and interpret - as a compensation of the way it was implemented compared to 

To me, it would make perfect sense to change the implementation to actually check the line separators _before_ tokens. It would cover more use cases whilst hardly adding any new code at all. True, it would add configuration, but I don't see that necessarily as a bad thing. At least it would address the concern of "addressing line separators between class members" in a single check, which is easier for users to understand.

If you put it in another check, there is a risk that both checks start diverging over time and working against each other, and then you have to take measures to compensate and such. So I really think keeping it one check is the right way forward.

In the meantime I've started working on a fix, I'd like the opportunity to at least show you what I have in mind.

**Comment by nmancus1** [comment]
> By changing the existing code, I can take the opportunity to make the existing code better.

> In the meantime I've started working on a fix, I'd like the opportunity to at least show you what I have in mind.

You are welcome to propose changes :) We can take a look at your idea, and go from there (regarding approval of issue).

Please be aware that quality bar is high if we even approve issue, we demand 100% code coverage, and 100% mutation coverage, all by real input (no pure unit tests).

**Comment by romani** [comment]
@peterdemaeyer , please share your expectations on class with two/few fields.

Your request looks reasonable, but we need more details to make sure we understand whole your style and we are not making problems to others.

**Comment by peterdemaeyer** [comment]
I already concluded that changing the existing check in a backwards compatible way is not possible, or at least not without adding even more complexity. I am not sure how far I am allowed to go in terms of incompatible changes, but I'm basically redefining the API of that whole check. Currently I'm defining a `StrictEmptyLineSeparatorCheck` as an alternative without touching `EmptyLineSeparatorCheck`, in order not to break anything, but in my mind it's a replacement.

The new API that I have in mind:

```
<module name="StrictEmptyLineSeparator">
    <!-- Below will be the default, which is similar to the existing EmptyLineSeparator, with the notable addition of RCURLY. -->
    <property name="tokens" value="PACKAGE_DEF, IMPORT, STATIC_IMPORT, CLASS_DEF, INTERFACE_DEF, ENUM_DEF, STATIC_INIT, INSTANCE_INIT, METHOD_DEF, CTOR_DEF, VARIABLE_DEF, RECORD_DEF, COMPACT_CTOR_DEF, RCURLY"/>
    <!-- Policies apply per token type, and use mini-expression language. -->
    <!-- The syntax is: "tokenType:minimumRequiredEmptyLines[:maximumAllowedEmptyLines][:lenientAfterLeftCurly][:coalesceTokensOfSameType]". -->
    <!-- Definitions apply to the empty lines BEFORE token type, and not AFTER. -->
    <property name="policies" values="PACKAGE_DEF=0:0, IMPORT=0:1, STATIC_IMPORT=0:1, RCURLY=0:0, *=1:1"/>
</module>
```

Let me first explain the basics "minimum required empty lines" and "maximum allowed empty lines". In the above example, the expression means:
- Before package definition, there are 0 empty lines allowed.
- Before import and static import, between 0 and 1 empty lines are allowed. Rely on the import checks to check any further restrictions, we don't need to be too rigorous here.
-- The range minRequired/maxAllowed replaces the old `allowMultipleEmptyLines` and allows finer-grained control.
- Before right curly, don't allow any empty lines.
- Before everything else, require and allow exactly one empty line (also after left curly).

Now let's explain what "lenient after left curly" and "coalesce tokens of same type" means.
If we were to express the default behaviour of the old `EmptyLineSeparator` using our new expression language, it would become:

```
<property name="tokens" value="all of the first example excluding RCURLY"/>
<property name="policies" value="PACKAGE_DEF=0:0, IMPORT=0:1, STATIC_IMPORT=0:1, VARIABLE_DEF=1:LENIENT:COALESCE, *=1:LENIENT"/>
```

The meaning of the above is:
- Exclude right curly. This disables any checking between class member and block end, because the old check didn't check that either.
- Before package definition, import and static imports, the meaning is the same as explained in the first example.
- Before variable definitions (field and static field definition), require at least one empty line, but be lenient after left curly (block start). The leniency here preserves backward compatibility. Also, allow coalescing variable definitions, but _only_ with other variable definitions, not between variable definitions and methods for example.
-- `coalesceTokensOfSameType` thus replaces the old `allowNoEmptyLinesBetweenFields`.
- Before everything else, require at least one empty line.

Just in case you don't like my expression language, I have an alternative in mind, which is to define the 4 properties of the check {minimumRequiredEmptyLines, maximumAllowedEmptyLines, lenientAfterLeftCurly, coalesceTokensOfSameType}, but that would require multiple instances of the check to be defined for different tokens. It would no longer be possible to have "one default configuration" for the check, the default configuration would consist of "multiple differently configured instances of the check". For that reason, I prefer the expression syntax approach, so that is what I am currently working on.

**Comment by romani** [comment]
@peterdemaeyer , thanks a lot for brain storm on this and sharing ideas with us.

> please share your expectations on class with two/few fields.

please reply.

**Comment by peterdemaeyer** [comment]
If I would have to change the existing check in a non-breaking way with minimal changes, I have two booleans in mind:
1. `checkBeforeInsteadOfAfter` which actually switches the check in "before" mode rather than "after". The default value would be `false` for backward compatibility. For my use case, it would be `true`.
2. `allowNoEmptyLinesBeforeBlockEnd` which flags violations if there are empty lines before block end (right curly). The default value would be `false` for backward compatibility. For my use case, it would be `true`.

I had started bolting these two onto the existing code, but it didn't feel like the most elegant solution and it was not so easy to keep the existing unit tests green, so I started trying a completely new alternative check, but after your comment I'll give the "minimal changes" solution another try.

**Comment by peterdemaeyer** [comment]
While attempting a fix, I encountered some unexpected behaviour: when traversing the AST, when finding an `RCURLY` in an empty block, the corresponding `LCURLY` is its _parent_while I expected it to be its _previous sibling_. For example:

```
class Test {
    static { // LCURLY
    } // RCURLY is the child of the correspong LCURLY and not its sibling???
}
```

**Comment by romani** [comment]
https://checkstyle.org/cmdline.html#Command_line_usage
Please use `-t` to show ast structure.
Our ast is not perfect by structure, but unfortunately if we missed inconsistency, it become set in stone.

**Comment by peterdemaeyer** [comment]
I created a PR for my "minimal non-breaking changes". There are 2 issues which I still need to resolve:
1. Translations to some languages. They make the build fail.
2. JaCoCo coverage is 96 %, I am missing some branches apparently, but IntelliJ shows me 100 % coverage and there is no HTML report as part of the build so I'm a bit blind as to what I am missing.

I could use some pointers for the remaining issues, but I already wanted to go ahead and create the PR to show you what I have in mind.

**Comment by nmancus1** [comment]
@peterdemaeyer please see https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-ut-coverage-report

**Comment by romani** [comment]
Please do not polish code until issue is approved, we do not want to waste your time.
We need to agree on general design and expectations first, there are a lot of nuances.

**Comment by peterdemaeyer** [comment]
Thanks for considering my time, but I spent it now anyway because I like to finish what I started, and amended my initial commit.
I won't blame you if you discard it, it was my own choice to spent my time.
Anyway, I fixed the translations as well as I could, and also the coverage is now at 100%.
The build now fails on Checkstyle violations that I can't fix without turning the whole class inside out.

```
[ERROR] [checkstyle] [ERROR] /home/peter/projects/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:346:1: Total number of methods is 38 (max allowed is 34). [MethodCount]
[ERROR] [checkstyle] [ERROR] /home/peter/projects/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:494:5: Cyclomatic Complexity is 12 (max allowed is 10). [CyclomaticComplexity]
[ERROR] [checkstyle] [ERROR] /home/peter/projects/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:494:5: NPath Complexity is 324 (max allowed is 200). [NPathComplexity]
```
I only added 4 methods or so, I'm quite sure the class was already at its limit before I started making changes.
In fact, it answers something I wondered about when fixing the issue: I had noticed that the `switch` block in `checkToken` nicely delegated to `processVariableDef/Import/Package` for all cases, _except_ for the default.
It made me wonder, why this inconsistency?
Now I know why... It would have been one method too much for the Checkstyle check, so the author inlined one arbitrary method to stay right under the limit.

This confirms that the existing check is running against its limits and it would be better to replace it.
Ideally I would like Checkstyle to be more configurable but stricter, in particular about the whitespace.
I'll probably end up writing my own custom `StrictEmptyLineSeparator` as a custom check, simply because I need it.

**Comment by romani** [comment]
>Checkstyle violations that I can't fix without turning the whole class inside out.

Put suppression on them, to bypass and see next problems, we will see how to fix it when we come to deep code review.

**Comment by peterdemaeyer** [comment]
> Put suppression on them

Done in an additional commit.

**Comment by peterdemaeyer** [comment]
In the meantime I've also completed my `StrictEmptyLineSeparatorCheck` (modulo the documentation), which has less than half of the LOC (and less general complexity) compared to the `EmptyLineSeparatorCheck`, and is more powerful. I haven't created a PR yet, but in my mind it would be the ultimate replacement for the current check. I've updated my comment of 20 Feb with the current implementation. I'll let you know when I create a PR for this.

**Comment by romani** [comment]
>I've also completed my StrictEmptyLineSeparatorCheck

If you want this Check be accessible to others, you can contribute it to https://github.com/sevntu-checkstyle/sevntu.checkstyle , we use this project as sandbox for Check to get maturity and let community to experiment.

**Comment by romani** [comment]
@peterdemaeyer , what do you think on proposal at https://github.com/checkstyle/checkstyle/issues/5439#issuecomment-1097041065 I think it will cover what you need and also it will be more customizable  for special team preferences.
